src/ipnd/services.rs

```diff
@@ -0,0 +1,287 @@
+extern crate alloc;
+use core::fmt;
+use log::info;
+use serde::de::{SeqAccess, Visitor};
+use serde::ser::{SerializeSeq, Serializer};
+use serde::{de, Deserialize, Deserializer, Serialize};
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+/// Struct representing the ServiceBlock used in Beacons to advertise additional services
+///
+/// Contains two vectors,
+///
+/// one reserved for ConvergencyLayerAgents
+///
+/// and one for user defined services
+#[derive(Debug, Clone, PartialEq)]
+pub struct ServiceBlock {
+    clas: Vec<(String, Option<u16>)>,
+    services: HashMap<u8, Vec<u8>>,
+}
+
+impl ServiceBlock {
+    /// Creates a new ServiceBlock without any services or clas
+    pub fn new() -> ServiceBlock {
+        ServiceBlock {
+            clas: Vec::new(),
+            services: HashMap::new(),
+        }
+    }
+
+    /// Returns the vector of ConvergencyLayerAgents
+    pub fn get_clas(&self) -> &Vec<(String, Option<u16>)> {
+        &self.clas
+    }
+
+    pub fn convert_clas(&self) -> Vec<String> {
+        let copy = self.clas.clone();
+        let mut convert = Vec::new();
+        for (scheme, port) in copy {
+            convert.push(format!("{}:{}", scheme, port.unwrap_or(0)));
+        }
+        convert
+    }
+
+    pub fn convert_services(&self) -> HashMap<u8, String> {
+        let mut convert: HashMap<u8, String> = HashMap::new();
+        for (tag, payload) in &self.services {
+            match tag {
+                63 => {
```

Comment:
these numbers are hard to remember. either constant u8s or enums as u8 should be used here